Pass trivially_copyable types with size < 16 bytes by val rather than by const ref&

jmarrec opened the pull request:
Pull request overview
---------------------

Apply a clang-tidy check I made: pass `trivially_copyable` types less than 16 bytes by val rather than by const ref&. It'll fit in a register and that's marginally faster.

see https://github.com/jmarrec/llvm-project/tree/misc-pod-const-ref for the clang-tidy check itself.

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

jmarrec committed:
Apply a clang-tidy check I made: pass trivially_copyable types less than 16 bytes by val rather than const ref&

see https://github.com/jmarrec/llvm-project/tree/misc-pod-const-ref

jmarrec committed:
Revert changes to SQLiteProcedures to avoid double precision problems

jmarrec commented on src/EnergyPlus/Autosizing/BaseSizerWithFanHeatInputs.cc:
```diff
@@ -57,13 +57,13 @@ namespace EnergyPlus {
     this->dataDesAccountForFanHeat = state.dataSize->DataDesAccountForFanHeat;
     // water coils on main branch have no parent object to set DataFan* variables
-    if (int(this->primaryAirSystem.size() > 0) && this->curSysNum > 0 && this->curOASysNum == 0) {
+    if (!(this->primaryAirSystem.empty()) && this->curSysNum > 0 && this->curOASysNum == 0) {
```
Manual change

amirroth commented:
I see what's going on now. Good stuff!

jmarrec commented on src/EnergyPlus/MixerComponent.cc:
```diff
@@ -626,32 +626,25 @@ void GetZoneMixerIndex(EnergyPlusData &state, std::string const &MixerName, int
-                if (InNodeNum != state.dataMixerComponent->MixerCond(MixerNum).InletNode(InNodeCtr)) continue;
-                thisMixer = MixerNum;
-                break;
+        for (int MixerNum = 1; MixerNum <= state.dataMixerComponent->NumMixers; ++MixerNum) {
+            for (int InNodeCtr = 1; InNodeCtr <= state.dataMixerComponent->MixerCond(MixerNum).NumInletNodes; ++InNodeCtr) {
+                if (InNodeNum == state.dataMixerComponent->MixerCond(MixerNum).InletNode(InNodeCtr)) {
+                    return MixerNum;
+                }
```
Manual change

jmarrec commented on src/EnergyPlus/ZonePlenum.hh:
```diff
@@ -182,11 +182,11 @@ namespace ZonePlenum {
+    int GetReturnPlenumIndex(EnergyPlusData &state, int ExNodeNum);
 
-    void GetReturnPlenumName(EnergyPlusData &state, int const &ReturnPlenumIndex, std::string &ReturnPlenumName);
+    void GetReturnPlenumName(EnergyPlusData &state, int ReturnPlenumIndex, std::string &ReturnPlenumName);
```
Should really be `std::string GetReturnPlenumName(EnergyPlusData &state, int ReturnPlenumIndex)`

amirroth replied:
Really, all of these objects should inherit from a single NamedObject class and use a single function `std::string GetObjectName(EnergyPlusData &state, Array1D<NameObject> &objects, int index)` rather than this proliferation of component specific lookup functions. I already opened an [issue](https://github.com/NREL/EnergyPlus/issues/8907) about this.

mitchute commented on src/EnergyPlus/BaseboardRadiator.cc:
```diff
@@ -1230,8 +1230,8 @@ namespace BaseboardRadiator {
     Real64 HWBaseboardUAResidual(EnergyPlusData &state,
-                                 Real64 const UA,                 // UA of coil
-                                 std::array<Real64, 2> const &Par // par(1) = design coil load [W]
+                                 Real64 const UA,                // UA of coil
+                                 std::array<Real64, 2> const Par // par(1) = design coil load [W]
```
I'm surprised by this one. Is this correct?

amirroth replied:
I'm surprised by this one too. From a performance standpoint, I think the argument is the same as the one for passing `std::string_view` and `gsl::span` by value. Each of those is a structure with two elements and passing two elements by value is a bit cheaper than passing the struct that contains them by reference (it's two instructions in the caller vs. one in the caller and two in the callee, so you are saving one instruction). However, despite this I do think it's better to use a reference in this case for the purpose of keeping the overall pattern simple, i.e., `std::string_view` and `gsl::span` by value and every other object by const reference. There is also the thing that `std::string_view` and `gsl::span` are themselves both essentially references and so passing a reference by reference is logically redundant, whereas that redundancy does not exist for random arrays of two elements.

jmarrec replied:
You have to remember that it's the first time I write a clang-tidy script, or even get exposed with the clang AST. I think I should have used a threshold of <= 15 bytes instead of <= 16 bytes when I applied it as well...

This is something I have done on my spare time (to address a specific OpenStudio issue that I could have addressed manually). I thought I'd try it out on E+ after mentioning it to Edwin, and I kinda messed this one up when applying to it took me like a half day to clean up manually, and I probably have missed a few things. This is just me nerdsnipping myself real bad (this isn't part of my scope so you won't find a pivotal ticket associated).

As far as the differences in output, I think some of the double/int arguments to the Serializing functions should be passed by const ref rather than value to avoid floating point errors/rounding...


I'm unfortunately short for free time to fix it for the next few weeks as I'm on paternity leave. I can close this issue or convert it to draft and we can revisit later?

mitchute replied:
Hey, I'm not knocking it - this is great! I've wanted to figure out how to use the clang AST tools for some time. Maybe you can give me a lesson when you're back 😉 .

Go be a dad and don't worry about this. I think one of us can clean up/review/merge this and get it wrapped up.

Congrats!

amirroth replied:
Congrats Julien! 🥳 👶

amirroth replied:
> You have to remember that it's the first time I write a clang-tidy script, or even get exposed with the clang AST. I think I should have used a threshold of <= 15 bytes instead of <= 16 bytes when I applied it as well...

The safe threshold is actually <= 8 bytes, but I'm not sure that looking only at argument size is enough. If you pass a struct/class with two 32-bit members or an array of 8 8-bit members by value, the elements would not be packed into a single register, they would be splayed out over multiple registers (I think). You wouldn't be able to access the members directly out of a single register. The safe check should be <= 8 bytes and not a class/struct or container. If it is not possible to implement that specific, then I think we need to do this by hand.

jmarrec replied:
Thanks!

I added two quick benchmarks to my benchmark (& testbed) repo here. See the commit here: https://github.com/jmarrec/CppBenchmarks/commit/2444a051ac51844c66bd41e3510f9e946610d322

It includes the results of running the benchmarks and the alrgorithmic complexity of each solution.

`const std::array<double, 2>&` is faster

mitchute reviewed (commented):
This looks fine. I am a little surprised by changing passing a ```std::array``` by value instead of reference, so please clarify if that's correct.

mitchute commented:
I pulled in develop locally and ran tests. Unit tests all pass serially. I also ran 50 regression tests and am seeing a few diffs. I expected this one to be totally diff-free so that probably needs to be resolved.

@jmarrec the files I'm seeing diffs on here are ASHRAE9012016_OfficeMedium_Denver and PythonPluginReplaceTraditionalManagers_LargeOffice.

Regarding staging of this PR: I'm fine holding this one since it touches a bunch of files. But at the same time, I doubt there would be any serious merge conflicts since all this does is touch the function declaration/definition signatures.

jmarrec commented on src/EnergyPlus/Autosizing/Base.cc:
```diff
@@ -623,7 +623,7 @@ void BaseSizer::overrideSizingString(std::string &string)
 }
 
-Real64 BaseSizer::setOAFracForZoneEqSizing(EnergyPlusData &state, Real64 const &desMassFlow, DataSizing::ZoneEqSizingData const &zoneEqSizing)
+Real64 BaseSizer::setOAFracForZoneEqSizing(EnergyPlusData &state, Real64 const desMassFlow, DataSizing::ZoneEqSizingData const &zoneEqSizing)
```
In this file, I'm not sure whether we want the `Real64` to pass by const ref to avoid floating precision errors. (the int/bool for sure we don't)

jmarrec commented on src/EnergyPlus/Autosizing/Base.cc:
```diff
@@ -61,7 +61,7 @@ namespace EnergyPlus {
                                    std::string_view const _compType,
                                    std::string_view const _compName,
-                                   bool const &_printWarningFlag,
+                                   bool _printWarningFlag,
```
Might want to actually keep the const here for clarity?

jmarrec committed:
Revert the std::array<Real64, 2> changes to keep byref (it's marginally faster)

nrel-bot-3 commented:
@jmarrec @lgentile it has been 28 days since this pull request was last updated.

Myoldmopar commented:
The diffs here _should_ just be due to being out of date with develop, right?  I'm inclined to try to clean this up and get it merged, knowing that we may want to follow-up to tweak some of the arguments soon after.  What does everyone think of that?  I guess the question really comes down to: "Is this PR 90% correct arguments and 10% that may need to be tweaked?" or some other split.  Anyway, if it turns out this should be pushed forward, I'll clean up conflicts.

amirroth commented:
I had some concerns about the actual search/replace logic here, but I think papa @jmarrec has addressed them.

jmarrec committed:
Merge remote-tracking branch 'origin/develop' into misc_pod_const_ref

jmarrec commented:
@Myoldmopar I merge develop in and (hopefully) resolved the conflicts correctly

Myoldmopar commented:
@jmarrec I see there is an issue here, and I'd be happy to try to resolve it, but I wanted to ask you about it.  CI is showing a couple things that appear to be related:

- Windows is showing a couple build warnings for uninitialized variables in the PhotoVoltaics module: https://nrel.github.io/EnergyPlusBuildResults/EnergyPlus-3d8dbfa8ec566260092cacadeb4158f840896c38-Win64-Windows-10-VisualStudio-16.html
- Windows is also joining with the other platforms in showing a failure in the GeneratorsWithPV test file.  Seems quite related.  

I assumed that when I looked at the files changed here, I would see that line being changed, or something related to those variables, but I don't.  So I'm not sure what's up.  If you have a chance to poke in there and just let me know if you notice anything, that would be great.  I'm happy to wrap this up.  If not I'll still take a crack at it soon.  Thanks!

jmarrec commented:
@Myoldmopar The signature of the NEWTON method changed: https://github.com/NREL/EnergyPlus/pull/9021/files#diff-8a7850a5bcef57a884f5544dc849c1a8b7b7a24957aeec1d9dfdc4183338cb88R1493

https://github.com/NREL/EnergyPlus/blob/3d8dbfa8ec566260092cacadeb4158f840896c38/src/EnergyPlus/Photovoltaics.cc#L1371

```
NEWTON(state, ISC, FUN, FI, ISC, DataPrecisionGlobals::constant_zero, IO, IL, SeriesResistance, AA, ISCG1, EPS);
                ^             ^                                                                         ^
        (Byref, output)      THAT GUY USED TO BE BYREF, NOW BYVAL                                       XS
```

ISC is indeed uninitialized when it's passed to the function on the first iteration.

Before it would be passed byref TWICE. The first arg is by ref and modified inside the function (and thus initialized inside the NEWTON method), to ISCG1 https://github.com/NREL/EnergyPlus/blob/658dce2e15c1c689281203426bf0f5cf48a13539/src/EnergyPlus/Photovoltaics.cc#L1518. That second ISC would therefore also be initialized by this line, but now it's by byval so it isn't.

jmarrec committed:
Don't bother trying to touch the NEWTON function

cf https://github.com/NREL/EnergyPlus/pull/9021#issuecomment-969321710 for why it's problematic to change it

jmarrec committed:
Merge remote-tracking branch 'origin/develop' into misc_pod_const_ref

amirroth commented:
The next person who copies something from "Numerical Recipes in Fortran" into EnergyPlus is fired.

Myoldmopar committed:
Merge develop; resolve conflicts

Myoldmopar commented:
Resolved conflicts and tested this out happily.  I think this could drop in if CI agrees.  Thanks @jmarrec and reviewers.

Myoldmopar commented:
Well that's interesting.  I didn't realize develop had broken just now due to a "string too big"!?  I'll fix that separately, but that shouldn't hold this up.  This looks good to go in, merging.  Thanks @jmarrec
